six/modules/python/cphd/source/cphd.i

```diff
@@ -143,6 +143,324 @@ using six::Vector3;
+
+        \return List of Python dictionaries containing NumPy arrays of PVP data.
+                Each dictionary in the list corresponds to a CPHD data channel.
+                The dictionary keys are string names of the PVP parameters in this PVPBlock
+                    (specifically, the names of the attributes used to store them in a CPHD PVP
+                    object, e.g. 'rcvTime').
+                The dictionary values are NumPy arrays of shape
+                    (cphdMetadata.getNumVectors(channel), cphdMetadata.getNumSamples(channel))
+                    (with an extra dimension of size cphdMetadata.pvp.[param].getSize() if the
+                    parameter size != 1).
+                The data types of these arrays are set based on the PVP format string,
+                    cphdMetadata.pvp.[param].getFormat(), using PVPBlock._pvpFormatToNPdtype()
+                Any added PVP parameters should also have been added to cphdMetadata.pvp.addedPVP
+        """
+
+        # getDefaultParametersInUse() maps all string param names to the names used in their
+        # get/set methods. Call it and reorganize a little: prepend 'get' and wrap the string
+        # method names in tuples
+        paramsToCopy = {paramName: ('get' + paramMethodName,)
+                        for paramName, paramMethodName in self.getDefaultParametersInUse().items()}
+
+        for paramName in paramsToCopy:
+            # Append PVPType object to tuple inside paramsToCopy
+            paramsToCopy[paramName] += (getattr(cphdMetadata.pvp, paramName),)
+        # Add custom PVP objects, which don't have getters
+        # (APVPType ("AddedPVPType") derives from PVPType)
+        paramsToCopy.update({paramName: (None, paramObj)
+                             for paramName, paramObj in cphdMetadata.pvp.addedPVP.items()})
+
+        # Now paramsToCopy consists of:
+        # {'param1Name': ('getParam1', cphd.PVPType object for param1)}
+        # for all default and custom PVP parameters used in this PVPBlock
+
+        pvpData = []
+        # Read data from each channel of this PVPBlock into list-of-dicts
+        for channel in range(cphdMetadata.getNumChannels()):
+            # Initialize dict of parameters for this channel
+            # Doing this for each channel in case they have different numbers of vectors/samples
+            channelPVP = {}
+            for paramName, (paramGetter, paramObj) in paramsToCopy.items():
+                paramSize = paramObj.getSize()
+                paramShape = (cphdMetadata.getNumVectors(channel),)
+                if paramSize != 1:
+                    # If data is a vector, add another dimension to the array
+                    paramShape += (paramSize,)
+                paramDtype = self._pvpFormatToNPdtype(paramObj.getFormat())
+                channelPVP[paramName] = numpy.empty(shape=paramShape, dtype=paramDtype)
+
+            # Copy PVP data for this channel by vector
```

Comment:
After looking more closely at this, I think we should strive to remove the loop over vectors. It can get very expensive very quickly. I already have an implementation that does this, but I haven't tested it on data containing `AddedPVP` stuff.

First get a mapping of the (valid) PVP data to their sizes, offsets, and formats (see comment above):
```
pvp_info = ...
```

Since the `PVPBlock` is just a giant binary blob in memory, we need to sort the `pvp_info` by offset:
```
from collections import OrderedDict

offsets = np.asarray([val[1] for val in pvp_info.values()])
offset_order = np.argsort(offsets)
sorted_keys = np.asarray(list(pvp_info.keys()))[offset_order]
sorted_pvp_info = OrderedDict([
    (key, pvp_info[key]) for key in sorted_keys
])
```

Next we need to extract the entire `PVPBlock` binary blob. Thankfully, there's an interface to do that:
```
num_vectors = metadata.getNumVectors(channel)
pvp_size = metadata.pvp.sizeInBytes()

pvp_data = np.array(
    num_vectors * pvp_size,
    dtype=np.bytes_
)

pvp_block.getPVPData(
    channel,
    pvp_data.ctypes.get_data(),
)

pvp_data = pvp_data.reshape(num_vectors, pvp_size)
```

The `pvp_data` array now has shape (num_vectors, pvp_bytes) with serialized byte entries. Getting at the per-entry data amounts to little more than slicing and viewing. For instance, if you know the first PVP entry (i.e. offset=0) is 8-byte floating point (i.e. format='F8'), then `pvp_data[:, :8]` will slice out those first 8 bytes across all vectors. To cast to `np.float64`, you'll have to do a copy since our slice is actually striding in memory:

```
pvp_entry = pvp_data[:, :8].copy().view(np.float64).squeeze()
```

The squeeze at the end ensures (N,1) arrays are returned as simple 1d arrays (N,). Putting this in a big loop give you:

```
pvp_dict = {}
pvp_offset = 0
for name, (offset, size, fmt) in sorted_pvp_info.items():
    # Split parameters into a format dict with keys equal to
    # param suffixes. Single parameters have entry suffix
    # Example: 'F8' --> {'': 'F8'}
    # Example: 'A=F8;B=I4' --> {'A': 'F8', 'B': 'I4'}
    suffix_to_format = _split_multi_format(fmt)

    # If a PVP has multiple parameters but all have the same
    # format, collapse them into a single PVP. This is intended
    # to simplify things like 'txPos', where it's more convenient
    # to have an (N,3)-shape array than three length-N arrays
    # for 'txPosX', 'txPosY', and 'txPosZ' separately.
    _size = 1
    formats = list(suffix_to_format.values())
    if _common_format(formats):
        suffix_to_format = {'': formats[0]}
        _size = size

    for suffix, _fmt in suffix_to_format.items():
        _name = name + suffix
        dtype = _format_to_dtype(_fmt)
        length = dtype.itemsize * _size
        vals = pvp_data[
            pvp_offset:(pvp_offset + length)
        ].copy().view(dtype)
        pvp_dict[_name] = vals.reshape(num_vectors, _size).squeeze()
        pvp_offset += length
```

A few of the details from the method above are:

```
def _common_format(formats):
    return len(set(formats)) == 1

def _split_multi_format(fmt):
    if ';' in fmt:
        result = {}
        for _fmt in fmt.split(';')[:fmt.count(';')]:
            assert '=' in _fmt
            name, __fmt = _fmt.split('=')
            result[name] = __fmt
        return result
    return {'': fmt}
```

Reply:
Why does `PVPBlock` set `signal` as a `double` [here](https://github.com/ngageoint/six-library/blob/master/six/modules/c%2B%2B/cphd/include/cphd/PVPBlock.h#L436) and in a few other places?  ('`I8'`= 8-byte signed integer) I think this might be related to an issue I'm having reading `signal` from the binary blob

Reply:
Was it because `double`s are 8 bytes and `int`s can be 4? (That's what they are on my machine at least).  Using `signed long long` ([which is guaranteed to be at least 8 bytes](https://en.cppreference.com/w/cpp/language/types)) instead of `double` for `signal` and added `'I8'` parameters fixes the issue I was having reading them from the binary blob.

Similarly, `std::complex<double>` is 16 bytes (8 real, 8 imaginary), which causes issues using it for custom `'CF8'` parameters, which are supposed to be 4 real and 4 imaginary bytes. `std::complex<double>` does work for `'CF16'` custom parameters though.

NumPy dtypes are flexible enough to handle different-sized parameters on the Python side, but I'm not sure what we can do in C++ to guarantee that the data being written is the correct size (and account for different C++ type sizes on different machines).

Reply:
There was a bug in the `signal` field initially. I think the author just copy-pasted `double` for the datatype and didn't notice it was different for `signal`. When the datatype was fixed from `F8` to `I8`, I'm not sure it was rippled to all other code (or the existing Python bindings) thoroughly. You're right it should be cast to `int64`

In terms of datatype compatibility between C++ and Python, I believe the following should work:
C++  <-->  Python
`double`  <-->  `np.float64`
`std::complex<float>`  <-->  `np.complex64`
`std::complex<double>`  <-->  `np.complex128`

The `sys` module in CODA-OSS offers abstractions for datatypes in Windows and non-Windows. Check out `Conf.h` if you want to see the explicit breakdown. For instance `sys::Int16_T` is a typedef for `int16_t` on Unix systems and `__int16` on Windows. There are no abstractions for floating point (I believe) because it assumes adherence to IEEE standards for things like `float`, `double`, etc. I think the IEEE standards also apply to `complex` in that the real and imaginary entries are side by side in memory.

Does that help at all?

Reply:
Yes, thank you, that helps!

Using `sys::Int64_T` for `signal` works correctly, and using other `sys` types for custom PVP parameters works better as well. In `PVPBlock.cpp`, each potential custom PVP parameter format string now corresponds to one of those `sys` types, rather than e.g. using `int` for all the `'I_'` parameters.

There are still a few issues when reading custom PVP parameters back into NumPy arrays:
 - `'F8'` and `'CF16'` (which both use 8-byte `double`s) are close (<`1e-09`), but not quite exact to what was passed in. I think this might be related to how custom PVP data is stored as `six::Parameter`s, which store data as strings, there might be some loss of precision there.
 - `'U1'` and `'I1'` parameters are both incorrect -- not yet sure why this is. The offset looks correct, and we only read one byte from the binary blob
 - NumPy has no complex integer type for `'CI_'` custom parameters. I will try something like [this](https://stackoverflow.com/a/32877245/3822302), but I'm not sure how it will work transferring to C++.

Reply:
I suspect the precision issue is related to what template instantiation of `AddedPVP<T>::getAddedPVP()` you're using. Looking at the top of `cphd.i`, it looks like there are only instantiations for `T` in {`unsigned`, `int`, `float`, `std::complex<int>`, `std::complex<float>`, and `std::string`}. Then, inside `pvp_format_to_added_pvp_methods`, the value of `method_name` is set to one of the explicit instantiations of the `getAddedPVP` getter. The value that `method_name` is set to only depends on the first entry of the format string.

For instance, suppose we're calling it on `"F8"` formatted data. We should be pulling out 8-byte, double-precision data, and we should be using `getAddedPVP<double>`. But instead we use `getAddedPVP<float>`. So we end up downcasting in the conversion and throwing away the extra precision below `O(10^-8)`. The same goes for `"CF16"` since we only provide template instantiation for `std::complex<float>` and not `std::complex<double>`. That probably explains the `10^-9` differences.

This probably also explains why `"U1"` and `"I1"` are incorrect. They're both using the template instantiation for `unsigned` and `int` respectively, both of which are 4 bytes here.

As for `"CI*"` formats, I suggest raising a `NotImplementedError` and calling it a day. There are ways around this using a structured array with a custom `dtype`, but we can always create an issue and circle back to that later. FWIW, I believe this would implement it for `"CI32"`:
```
ci32_dtype = np.dtype([
    ('real', np.int16),
    ('imag', np.int16),
])
data = np.empty(shape, dtype=ci32_dtype)
```

Reply:
Yes, that fixes the float precision errors, thank you!

I'm still having issues with `'U1'` and `'I1'`. It looks like `sys::Int8_T` is a typedef for `int8_t`, which in turn is a typedef for `signed char` (which is what SWIG expects `setI1AddedPVP()` to receive). In `from_list_of_dicts()`, we're passing it a Python `int` (converted from a `np.int8`), so I wonder if there might be some hidden type conversion going on here. I tried using `struct.pack('b', param_data)` and converting `param_data` to a `ctypes.c_byte`, but the `setI1AddedPVP()` wouldn't accept either, saying that they aren't `signed char`s.

`NotImplementedError` sounds good for `'CI*'` formats. I'll also raise that for custom PVP parameters with multiple formats. I did try using structured arrays and custom NumPy `dtype`s , and I think we should be able to read them from the binary blob more or less as-is, we just aren't able to set them without accessing the binary blob.